### PR TITLE
Add brute force protection to login screen

### PR DESCRIPTION
**Pull request by saket**

### CODE REVIEW

**Comment by vinaysshenoy** [comment]
@saket What do you feel about squashing b837371749c05d0f39c2f5248e0acf776713dead and 99c13dee44cca8e877c2e91d15e802c2877ae306?

**Comment by saket** [comment]
@vinaysshenoy sure I'll squash them.

File: app/src/test/java/org/simple/clinic/security/pin/BruteForceProtectionStateTest.kt [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+import org.junit.Test
+import org.simple.clinic.di.NetworkModule
+import org.simple.clinic.util.Just
+import org.threeten.bp.Instant
+
+class BruteForceProtectionStateTest {
+
+  /**
+   * If you're seeing this test fail, it means that you modified
+   * [BruteForceProtectionState] without thinking about migration.
```
Comment (vinaysshenoy) [inline app/src/test/java/org/simple/clinic/security/pin/BruteForceProtectionStateTest.kt]: We also need to plan as to how exactly we **do** migration for shared preferences, if ever required.

File: app/src/main/java/org/simple/clinic/security/pin/PinEntryCardView.kt [hunk 2]
```diff
@@ -98,11 +100,11 @@ class PinEntryCardView(context: Context, attrs: AttributeSet) : CardView(context
       is State.Progress -> VISIBLE
     }
 
-    // TODO: Use IDs instead of indices.
-    pinAndLockViewFlipper.displayedChild = when (state) {
-      State.PinEntry -> 0
-      State.Progress -> pinAndLockViewFlipper.displayedChild
-      is State.BruteForceLocked -> 1
+    Timber.i("Moving to state: $state")
```
Comment (vinaysshenoy) [inline app/src/main/java/org/simple/clinic/security/pin/PinEntryCardView.kt]: Leftover log?

File: app/src/main/java/org/simple/clinic/widgets/Views.kt [hunk 3]
```diff
@@ -165,3 +166,13 @@ fun TextView.setTextAppearanceCompat(@StyleRes resourceId: Int) {
...
   }
 }
+
+var ViewFlipper.displayedChildResId: Int
+  @IdRes
+  get() = getChildAt(displayedChild).id
+  set(@IdRes id) {
+    val child = findViewById<View>(id)
+    val childIndex = indexOfChild(child)
+    Timber.i("Setting to $displayedChild")
```
Comment (vinaysshenoy) [inline app/src/main/java/org/simple/clinic/widgets/Views.kt]: Leftover log?

File: app/src/main/java/org/simple/clinic/util/OptionalMoshiAdapter.kt [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+class MoshiOptionalAdapterFactory : JsonAdapter.Factory {
+
+  override fun create(type: Type, annotations: Set<Annotation>, moshi: Moshi): JsonAdapter<*>? {
+    return if (type is ParameterizedType && type.rawType === Optional::class.java) {
+      OptionalAdapter(moshi, type.actualTypeArguments[0])
+    } else null
+  }
+
+  private class OptionalAdapter internal constructor(private val moshi: Moshi, private val type: Type) : JsonAdapter<Optional<*>>() {
```
[the remark this answers is not in the document]
Comment (saket) [reply app/src/main/java/org/simple/clinic/util/OptionalMoshiAdapter.kt]: Whoops. The Kotlin code converter generated this automatically. I'll remove this.

File: app/src/main/java/org/simple/clinic/util/OptionalMoshiAdapter.kt [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    @Throws(IOException::class)
+    override fun toJson(out: JsonWriter, value: Optional<*>?) {
+      when (value) {
+        is Just -> moshi.adapter<Any>(type).toJson(out, value.value)
+        is None -> out.nullValue()
+      }
+    }
+
+    @Throws(IOException::class)
+    override fun fromJson(`in`: JsonReader): Optional<*>? {
```
[the remark this answers is not in the document]
Comment (saket) [reply app/src/main/java/org/simple/clinic/util/OptionalMoshiAdapter.kt]: Yes, will do this.

File: app/src/main/java/org/simple/clinic/login/pin/LoginPinScreenController.kt [hunk 6]
```diff
@@ -42,63 +36,29 @@ class LoginPinScreenController @Inject constructor(
   }
 
   private fun submitClicks(events: Observable<UiEvent>): Observable<UiChange> {
-    val pinChanges = events.ofType<PinTextChanged>()
+    return events.ofType<LoginPinAuthenticated>()
         .map { it.pin }
+        .flatMapSingle { pin ->
+          Timber.i("Requesting OTP")
```
[the remark this answers is not in the document]
Comment (saket) [reply app/src/main/java/org/simple/clinic/login/pin/LoginPinScreenController.kt]: Removed

File: app/src/main/java/org/simple/clinic/security/pin/PinEntryCardView.kt (hunk 2 above)
Comment (saket) [reply app/src/main/java/org/simple/clinic/security/pin/PinEntryCardView.kt]: Removed

File: app/src/main/java/org/simple/clinic/widgets/Views.kt (hunk 3 above)
Comment (saket) [reply app/src/main/java/org/simple/clinic/widgets/Views.kt]: Removed

File: app/src/test/java/org/simple/clinic/security/pin/BruteForceProtectionStateTest.kt (hunk 1 above)
Comment (saket) [reply app/src/test/java/org/simple/clinic/security/pin/BruteForceProtectionStateTest.kt]: Done

**Pull request merged**